### PR TITLE
fix(sbom): attach nested packages to Application

### LINKED ISSUE #8143: bug(sbom): multiple Results for one package with root relationship for `sbom` mode
## Description
For `sbom` mode Trivy creates 2 Results for one Application (if Package has Root relationship)
e.g.:
```
➜  ./trivy -q rootfs -f cyclonedx -o report.cdx.json ./pkg/dependency/parser/golang/binary/testdata/test.elf
➜  ./trivy -q sbom -f json --list-all-pkgs  ./report.cdx.json | grep Target -A 2                            
      "Target": "",
      "Class": "lang-pkgs",
      "Type": "gobinary",
--
      "Target": "test.elf",
      "Class": "lang-pkgs",
      "Type": "gobinary",
```

### PR DESCRIPTION
**Pull request by DmitriyLewen**
## Description
There is a problem with decoding nested packages.
We create 2 attachments for them (see #8143).

This PR recursively checks all related packages and associates them with an attachment.

Example ([test file packages](https://github.com/aquasecurity/trivy/blob/bcdc0bbf1f63777ff79d3ecadb8d4f916f376b7d/pkg/dependency/parser/golang/binary/parse_test.go#L15-L54):
Before:
```bash
➜ trivy -q rootfs -f cyclonedx -o report.cdx.json ./pkg/dependency/parser/golang/binary/testdata/test.elf && trivy -q sbom ./report.cdx.json -f json --list-all-pkgs | grep Target 
      "Target": "",
      "Target": "test.elf",
```

After:
```bash
➜ trivy -q rootfs -f cyclonedx -o report.cdx.json ./pkg/dependency/parser/golang/binary/testdata/test.elf && ./trivy -q sbom ./report.cdx.json -f json --list-all-pkgs | grep Target
      "Target": "test.elf",
```

## Related issues
- Close #8143

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [ ] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [ ] I've added usage information (if the PR introduces new options)
- [x] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

File: pkg/sbom/cyclonedx/testdata/happy/nested-packages-bom.json [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+      ]
+    },
+    {
+      "ref": "4021d631-e242-4e69-8a93-928665810a27",
+      "dependsOn": [
+        "pkg:golang/github.com/aquasecurity/go-pep440-version@v0.0.0-20210121094942-22b2f8951d46",
+        "pkg:golang/github.com/aquasecurity/go-version@v0.0.0-20210121072130-637058cfe492",
+        "pkg:golang/github.com/aquasecurity/test",
+        "pkg:golang/golang.org/x/xerrors@v0.0.0-20200804184101-5ec99f83aff1"
+      ]
```
Comment (DmitriyLewen) [inline pkg/sbom/cyclonedx/testdata/happy/nested-packages-bom.json:146]: These dependencies will be changed after #8104

So this should be changed in one of these PRs

Comment (DmitriyLewen) [reply pkg/sbom/cyclonedx/testdata/happy/nested-packages-bom.json]: updated in https://github.com/aquasecurity/trivy/pull/8144/commits/40d5dac6f0891aea3acfad955e5059e2e14ac28d

**Comment by knqyf263** [comment]
@DmitriyLewen I've added three changes to prevent potential issues:

1. Recursive Processing for OS Packages:
   - Added recursive processing for OS packages to prevent future bugs
   - This preemptive change ensures consistency with language-specific package handling
   - Without this change, future modifications to OS package handling could lead to incomplete dependency resolution

2. Infinite Loop Prevention:
   - Implemented cycle detection using a visited packages map
   - Each package is now processed only once to prevent infinite recursion

3. Package Duplication Resolution:
   - Previous versions of Trivy could output SBOMs where the same package appears multiple times in the dependency tree
   - This was due to packages with unknown relationships being added under applications (This behavior was changed in https://github.com/aquasecurity/trivy/pull/8104)
   - The cycle detection mechanism (change #2) also resolves this package duplication issue

Please verify these changes maintain the expected behavior while addressing the identified issues.

File: pkg/sbom/io/decode.go [hunk 2]
```diff
@@ -348,18 +341,37 @@ func (m *Decoder) addOSPkgs(sbom *types.SBOM) {
...
+// It starts from the given component ID and traverses the dependency tree, collecting all
+// dependent packages. The collected packages are removed from m.pkgs to prevent duplicate
+// processing. This ensures that all dependencies, including transitive ones, are properly
+// captured and associated with their parent component.
+func (m *Decoder) traverseDependencies(id uuid.UUID, seen set.Set[uuid.UUID]) ftypes.Packages {
+	var pkgs ftypes.Packages
+	for _, rel := range m.bom.Relationships()[id] {
+		if seen.Contains(rel.Dependency) {
+			continue // Skip if already seen to avoid infinite loop and duplicate processing
+		}
```
Comment (DmitriyLewen) [inline pkg/sbom/io/decode.go:359]: I checked case with infinity loop.
We delete package from `m.pkgs`:
https://github.com/aquasecurity/trivy/blob/b9e85e731173ce981ce93573d31e82f3fb4d0525/pkg/sbom/io/decode.go#L367

Therefore, the following logic will avoid infinite loops:
https://github.com/aquasecurity/trivy/blob/b9e85e731173ce981ce93573d31e82f3fb4d0525/pkg/sbom/io/decode.go#L360-L363

Or am i missing something?

Comment (knqyf263) [reply pkg/sbom/io/decode.go]: Oh, you're right. I was smarter in the past 😄 Reverted https://github.com/aquasecurity/trivy/pull/8144/commits/4db29eb054a7d8291bf9914b0827bd8e2572c265

**Comment by DmitriyLewen** [comment]
Thank you for handling the OS packages.
I didn't think about them.

Looks good now 👍

> Please verify these changes maintain the expected behavior while addressing the identified issues.

It works as expected.
Approve the PR if you have no other changes.

**Review by knqyf263 — APPROVED** [review]

**Comment by DmitriyLewen** [comment]
@aqua-bot backport release/v0.58

**Pull request merged**